### PR TITLE
Used glide to vendor dependencies

### PR DESCRIPTION
**Pull request by zefciu**
In order to introduce glog (and possibly other dependencies) we need
a flat vendor directory (glog panics if two versions of it are
imported). We decided to use glide. It seems to be developer-friendly
and also offers mechanism for consistend deploys

Some questions you may have:

Q: What with the size of repo
A: We don't store any additional code in repo. Only the hashes in
   glide.lock

Q: How can I use my own k8s version.
A: Edit the glide.lock and put your commit hash there

Q: The difference between update and install
A: glide update will modify glide.lock. glide install only installs
   stuff

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-appcontroller/93)

### CODE REVIEW

**Comment by jellonek** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[Dockerfile, line 17 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTK4HcQP-kle7kY7I1o:-KTK4HcQP-kle7kY7I1p:be1hziw) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/c081cf6ad0f6d817655f9102cc49c1f4816f31c6/Dockerfile#L17)):*
> ```Dockerfile
>     mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests &&\
>     rm -fr /go
> ```

Unnecessary empty line.

---

*[README.md, line 120 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTK3a6ZqMJ1QRsDXJ7d:-KTK3a6ZqMJ1QRsDXJ7e:be1hzkt) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/c081cf6ad0f6d817655f9102cc49c1f4816f31c6/README.md#L120)):*
> ```Markdown
> 
> You can also override this behaviour by using `-l` flag for `kubeac deploy` command available on AppController pod, but this should be only done for testing purposes and is not encouraged in production.
> ```

Unnecessary empty lines

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93)*

**Comment by jellonek** [comment]
Looks like travis seeing `Makefile` uses `make` to build project, before running tests. Maybe you should add `script` entry there as described in https://docs.travis-ci.com/user/languages/go/ ?

**Comment by jellonek** [comment]
Hmmm... What is the reason behind using `glide-vc` if we are not vendoring anything in project?
IMO using it makes sense only when `vendor/` is included in project, to minimize it's size to only necessary part of dependent files.

**Comment by pigmej** [comment]
@jellonek smaller binary size, smaller file with deps. Isn't it?

**Comment by jellonek** [comment]
Didn't noticed binary size change (have to check - possibly by eliminating files not used by our source but having `init` func).
Smaller file with deps? Which file? repo tgz?

**Comment by gluke77** [comment]
I'd second jellonek as for getting rid of glide-vc

**Comment by pigmej** [comment]
If it changes nothing then let's get rid of it. :)

**Comment by jellonek** [comment]
Only change it now provides is increasing time of build :P

**Comment by zefciu** [comment]
Testing travis manifest

File: Makefile [hunk 1]
```diff
@@ -10,4 +10,4 @@ vendor: Makefile
 	glide-vc --only-code --no-tests
 
 test: vendor glide.lock Makefile
```
Comment (jellonek) [inline Makefile]: Remove `Makefile` from deps.

File: glide.lock [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  - pkg/util/wait
+  - pkg/util/yaml
+  - pkg/version
+  - pkg/watch
+  - pkg/watch/versioned
+  - plugin/pkg/client/auth
+  - plugin/pkg/client/auth/gcp
+  - plugin/pkg/client/auth/oidc
+  - third_party/forked/golang/reflect
+  - vendor/github.com/golang/glog
```
Comment (jellonek) [inline glide.lock]: This is really strange. IMO this should not be included if there is no call to it from appcontroller.

File: Makefile (hunk 1 above)
Comment (zefciu) [inline Makefile]: Why?

Comment (jellonek) [inline Makefile]: Misspointed question. It's ok.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+TAG ?= mirantis/k8s-appcontroller
+
+docker: vendor glide.lock Makefile
+	docker build -t $(TAG) .
+
+vendor: Makefile
+	glide install --strip-vendor
+
+test: vendor glide.lock Makefile
```
Comment (zefciu) [inline Makefile]: Why? You can't run tests if the 'vendor' directory is not created.

File: glide.lock (hunk 2 above)
Comment (zefciu) [inline glide.lock]: It seems that glide.lock must contain all the flattened dependencies

**Comment by zefciu** [comment]
Review status: 1 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[glide.lock, line 317 at r5](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ6-r5-317:-KTheCuFpWUxoWgfLWCk:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/bd3ec117afd725957682557c6d984a0238ea469b/glide.lock#L317)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> It seems that glide.lock must contain all the flattened dependencies
</details>
Done.

---

*[Makefile, line 12 at r4](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ5-r4-12:-KTheDzFotV0swca-sla:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f5d69753c2c07b9138b6ecdabda015148df82f32/Makefile#L12)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary>
> Misspointed question. It's ok.
</details>
Done.

---

*[Makefile, line 11 at r5](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ5-r5-11:-KTheMQMcEP67e2vDYHe:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/bd3ec117afd725957682557c6d984a0238ea469b/Makefile#L11)):*
<details><summary><i>Previously, zefciu (Szymon Pyżalski) wrote…</i></summary>
> Why? You can't run tests if the 'vendor' directory is not created.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93)*

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,12 @@
+.PHONY: docker
+
+TAG ?= mirantis/k8s-appcontroller
+
+docker: vendor glide.lock Makefile
```
Comment (nebril) [inline Makefile]: should glide.lock be here? it seems like it's already handled in dockerfile

**Comment by nebril** [comment]
The image seems to be really big. 700MB is much more than was before: https://hub.docker.com/r/zefciu/k8s-appcontroller/builds/bnvvkgpchasbhpmrfba2hz3/

**Comment by jellonek** [comment]
Image size problem seems to be resolved. 👍

File: Dockerfile [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
 
 WORKDIR /go/src/github.com/Mirantis/k8s-AppController
 
-RUN apk --no-cache add git
-RUN go get ./... && go build -o kubeac && mv kubeac /usr/bin/kubeac && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
 
-RUN echo "@community http://dl-cdn.alpinelinux.org/alpine/edge/community" >> /etc/apk/repositories
-RUN apk --no-cache add runit@community
+RUN apk --no-cache add git runit@community glide@community &&\
+    glide i --strip-vendor &&\
```
Comment (gluke77) [inline Dockerfile]: Please do not use aliases for glide command

File: Dockerfile [hunk 6]
```diff
@@ -1,15 +1,23 @@
...
+    go build -o kubeac &&\
+    mv kubeac /usr/bin/kubeac &&\
+    mkdir -p /opt/kubeac &&\
+    mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests &&\
+    glide cc &&\
```
Comment (gluke77) [inline Dockerfile]: Ditto

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+TAG ?= mirantis/k8s-appcontroller
+
+docker: vendor Makefile
+	docker build -t $(TAG) .
+
+vendor: Makefile
+	glide install --strip-vendor
+
+test: vendor glide.lock Makefile
+	go list ./... | grep -v /vendor/ | xargs go test
```
Comment (gluke77) [inline Makefile]: Please use 'go test -run ...' to filter tests

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,12 @@
+.PHONY: docker
+
+TAG ?= mirantis/k8s-appcontroller
+
+docker: vendor Makefile
+	docker build -t $(TAG) .
+
+vendor: Makefile
+	glide install --strip-vendor
```
Comment (gluke77) [inline Makefile]: So we run 'glide install'  twice while building a docker img: here and in the dockerfile. THat does not look right. We probably should do something with it.

**Comment by zefciu** [comment]
Resolved the problem.

---

Review status: 1 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[Dockerfile, line 12 at r7](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ2-r7-12:-KTmkcBSnOZc2G5yQjiX:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/d04b7c85fd37b9d46cdc390ab4e5d704ecbe45b4/Dockerfile#L12)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> Please do not use aliases for glide command
</details>
Done.

---

*[Dockerfile, line 17 at r7](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ2-r7-17:-KTmkchKvjJwMsnklwm4:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/d04b7c85fd37b9d46cdc390ab4e5d704ecbe45b4/Dockerfile#L17)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> Ditto
</details>
Done.

---

*[Makefile, line 5 at r6](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ5-r6-5:-KTmkjBgG69-gajTkkMt:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/97a92507b2e4ab88225d98f91b47d752610622ef/Makefile#L5)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> should glide.lock be here? it seems like it's already handled in dockerfile
</details>
Done.

---

*[Makefile, line 9 at r7](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ5-r7-9:-KTmkknhdJTU5WlYrJnE:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/d04b7c85fd37b9d46cdc390ab4e5d704ecbe45b4/Makefile#L9)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> So we run 'glide install'  twice while building a docker img: here and in the dockerfile. THat does not look right. We probably should do something with it.
</details>
Done.

---

*[Makefile, line 12 at r7](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-KTJuJhQ1lXQ36bb6EZ5-r7-12:-KTmnXu44FTQU_xnj1ZJ:b70t5on) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/d04b7c85fd37b9d46cdc390ab4e5d704ecbe45b4/Makefile#L12)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> Please use 'go test -run ...' to filter tests
</details>
Can I do it without negative lookaheads?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-:-KTmkNRwlBKQ-F3A3__d:b-ez8l9b)*

**Comment by zefciu** [comment]
@gluke77 @nebril ptal

**Comment by jellonek** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r5, 2 of 3 files at r6, 2 of 3 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-:-KTnMQFork6ZQJ-5l8PG:bnfp4nl)*

File: Makefile (hunk 8 above)
Comment (zefciu) [inline Makefile]: Done

**Comment by gluke77** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/93#-:-KTnQiIQItTQ-8tzeSqc:bnfp4nl)*

**Pull request merged**